[DRAFT] Update mapping application architecture

rbsec opened the pull request:
**DRAFT - not ready for merging.**

Initial skeleton (with some of the old text) for a rewrite of the application architecture mapping guide. Initial thoughts are very welcome in terms of:

- Any major areas you think are missing?
- Any interesting ideas for identifying specific components (especially PaaS/serverless)?
- Should microservices be added? Is there any good way to detect them?

ThunderSon commented:
Looking like a good first round.
I highly believe they have their own headers. I might play around in a couple of weeks with them, and I'll see if I can summon someone that have worked with them previously.

Microservices are usually identified by the different responses (different headers, different error handling, etc.) It's very rare that a microservice is composed of the same language, same web server, same APIG, and same input/output logic.

rbsec commented:
@kingthorin and @ThunderSon - I think that this is ready for a *technical* review. I've not gone through to tidy up spelling/grammar/etc - but if you have thoughts on the content then please let me know. Once we're happy with that I'll give it a proper proof read.

ThunderSon commented on document/4-Web_Application_Security_Testing/01-Information_Gathering/10-Map_Application_Architecture.md:
```diff
@@ -6,39 +6,185 @@
+
+It is not always possible to detect a reverse proxy (especially if there is only a application behind it), but you can often sometimes identify it by:
+
+- A mismatch between the front end server and the back end application (such as a `Server: nginx` header with an ASP.NET application).
```
We can mention request smuggling as well (there is a test in the WSTG, but it's outdated now), so portswigger's documentation should do the trick till it's updated. Mainly referencing the headers and the mismatch that occurs between the 2 systems at play.

ThunderSon commented on document/4-Web_Application_Security_Testing/01-Information_Gathering/10-Map_Application_Architecture.md:
```diff
@@ -6,39 +6,185 @@
+
+It is typically configured by pointing the publicly facing domain to the CDN's servers, and then configuring the CDN to connect to the correct back end servers (sometimes known as the "origin").
+
+The easiest way to detect a CDN is to perform a WHOIS lookup for the IP addresses that the domain resolves to. If they belong to a CDN company (such as Akami, Cloudflare or Fastly - see [Wikipedia](https://en.wikipedia.org/wiki/Content_delivery_network#Notable_content_delivery_service_providers) for a more complete list) then it's like that a CDN is in use.
```
Just fixed Akamai typo, might go under the radar
```suggestion
The easiest way to detect a CDN is to perform a WHOIS lookup for the IP addresses that the domain resolves to. If they belong to a CDN company (such as Akamai, Cloudflare or Fastly - see [Wikipedia](https://en.wikipedia.org/wiki/Content_delivery_network#Notable_content_delivery_service_providers) for a more complete list) then it's like that a CDN is in use.
```

ThunderSon commented on document/4-Web_Application_Security_Testing/01-Information_Gathering/10-Map_Application_Architecture.md:
```suggestion
A network Intrusion Detection System (IDS) detects suspicious or malicious network-level activity (such as port or vulnerability scanning) and raises alerts. An Intrusion Prevention System (IPS) is similar, but also takes action to prevent the activity - usually by blocking the source IP address.
```

ThunderSon commented on document/4-Web_Application_Security_Testing/01-Information_Gathering/10-Map_Application_Architecture.md:
```suggestion
A Web Application Firewall (WAF) inspects the contents of HTTP requests and blocks ones that appear to be suspicious or malicious. They are usually based on a set of known bad signatures and regular expressions, such as the [OWASP Core Rule Set](https://owasp.org/www-project-modsecurity-core-rule-set/).  WAFs can be effective at protecting against some types of attacks (such as SQL injection or cross-site scripting), but are less effective against other types (such as access control or business logic related issues).
```

ThunderSon commented on document/4-Web_Application_Security_Testing/01-Information_Gathering/10-Map_Application_Architecture.md:
If WAFs are configured to handle states as well, they will inject their own cookies or headers to keep track of the user's activity, they might as well have different responses based on the behaviour. Like adding a bot captcha in occasions, and in other times having a simpler flow.

ThunderSon reviewed (changes requested):
I'd say "Frameworks" can make an appearance as well. What do you think?
I love the way it's looking, most comments aren't major.

You didn't go deep into topics, but gave the required information to map it out.

rbsec commented on document/4-Web_Application_Security_Testing/01-Information_Gathering/10-Map_Application_Architecture.md:
```diff
@@ -6,39 +6,185 @@
+
+Almost all web applications include third party resources that are loaded or interacted with by the client. These can include:
+
+- Active content (such as scripts, style sheets, fonts and iframes).
+- Passive content (such as images and videos).
```
I think that the examples here make it fairly clear what they mean, and it's a common term (mostly seen in the "mixed active content" warning browsers give):

https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content

ThunderSon replied:
Awesome we can link inline to the passive and active sections. I didn't bother searching for links when I was reviewing. I am saying that in the WSTG, at this stage, I didn't think we covered these types of contents.

rbsec commented:
@ThunderSon I've addressed most of your comments in b6f707b

With regards to frameworks, there's a whole document on them already in the information gathering section - so I'm not sure it's really worth adding a bit in here that just says "A framework may be used, read the guide on it" (especially as they'll probably have just read it earlier).

ThunderSon reviewed (approved):
The checklist will be auto-updated, no need to change it manually.
For the content, I am happy with it, would love to add inline the passive and active direct link to the sections in the MDN link you provided.
Not sure we want to link just yet to the request smuggling testing. Portswigger's documentation is much better on that topic.

rbsec commented:
@ThunderSon I've added the MDN links, and changes the smuggling one to PortSwigger. Probably worth putting that on the todo list for WSTG, but for now we're pointing at a better reference, so that's probably fine.

I'll have a proper read through tomorrow and clean up any spelling/grammar/etc.

kingthorin commented:
FYI this is still on my todo list, might not get to it till the weekend or even early next week.
